x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java

```diff
@@ -393,13 +463,30 @@ void waitForDatafeedStopped(List<String> datafeedPersistentTaskIds, StopDatafeed
                 throw org.elasticsearch.ExceptionsHelper
                         .convertToElastic(failedNodeExceptions.get(0));
             } else {
-                // This can happen when the actual task in the node no longer exists,
-                // which means the datafeed(s) have already been stopped.
-                return new StopDatafeedAction.Response(true);
```

Comment:
We were discarding significant information here.  This is where we could have detected the case of "persistent task implies local task is running on node X, no local task currently running on node X".

When we get here we don't know if the reason is that the local task hasn't yet started or that the local task has finished but the persistent task hasn't yet been removed.  In the first case we have to do something.  In the second we don't, but it doesn't hurt to retry.

I didn't add similar logic to the close job code.  For some reason we never seem to hit the problematic case with jobs.  This may be because job state has the concept of being stale, so we can detect recently relocated jobs by their staleness.  I don't want to add more complex code unnecessarily, but if we ever see the case of jobs closures timing out then something like what's now below may be the solution.